src/bidiMapper/modules/context/NavigationTracker.ts

```diff
@@ -0,0 +1,214 @@
+      this.#currentNavigation.markNavigationAborted();
+    }
+    const navigation = new NavigationState(this.#browsingContextId, url);
+    this.#setEventListeners(navigation);
+
+    this.#currentNavigation = navigation;
+    return navigation;
+  }
+
+  #setEventListeners(navigation: NavigationState) {
+    void navigation.started.then(() => {
+      this.#eventManager.registerEvent(
+        {
+          type: 'event',
+          method: ChromiumBidi.BrowsingContext.EventNames.NavigationStarted,
+          params: navigation.navigationInfo(),
+        },
+        this.#browsingContextId,
+      );
+      return;
+    });
+
+    void navigation.finished.then((eventName: NavigationEventName) => {
+      this.#eventManager.registerEvent(
+        {
+          type: 'event',
+          method: eventName,
+          params: navigation.navigationInfo(),
+        },
+        this.#browsingContextId,
+      );
+      return;
+    });
+  }
+
+  frameNavigated(url: string): void {
+    this.#currentNavigation.url = url;
+  }
+
+  navigatedWithinDocument(url: string, navigationType: string): void {
+    this.#currentNavigation.url = url;
+    if (navigationType === 'fragment') {
+      this.#currentNavigation.markFragmentNavigated();
+    }
+  }
+
+  frameStartedLoading(): void {
+    this.#currentNavigation.markStarted();
+    this.#currentNavigation.cdpStates.frameStartedLoading = true;
+  }
+
+  frameScheduledNavigation(url: string): void {
+    // Signals that it's a new navigation:
+    // 1. The `Page.frameStartedLoading` was already emitted.
+    // 2. The `Page.frameScheduledNavigation` was already emitted.
+    // 3. The `Page.frameRequestedNavigation` was already emitted, which cannot happen
+    //    before `Page.frameScheduledNavigation`.
+    // 4. The current navigation is the initial one, and the URL does not match
+    //    `about:blank`.
+    const isNewNavigation =
+      this.#currentNavigation.cdpStates.frameStartedLoading ||
+      this.#currentNavigation.cdpStates.frameRequestedNavigation ||
+      this.#currentNavigation.cdpStates.frameScheduledNavigation ||
+      (this.#initialNavigation && !urlMatchesAboutBlank(url));
+
+    if (isNewNavigation) {
+      this.createNavigation(url);
+    } else {
+      this.#currentNavigation.url = url;
+    }
+    this.#currentNavigation.markStarted();
+    this.#currentNavigation.cdpStates.frameScheduledNavigation = true;
+  }
+
+  frameRequestedNavigation = (url: string): void => {
+    // Signals that it's a new navigation:
+    // 1. The `Page.frameStartedLoading` was already emitted.
+    // 2. The `Page.frameRequestedNavigation` was already emitted. Note that having
+    //    `Page.frameScheduledNavigation` emitted at this point does not signal the new
+    //    navigation.
+    // 3. The current navigation is the initial one, and the URL does not match
+    //    `about:blank`.
+    const isNewNavigation =
+      this.#currentNavigation.cdpStates.frameStartedLoading ||
```

Comment:
how does cdpStates track if the events came from a current navigation or from one of the past navigations?